### PR TITLE
Legends

### LINKED ISSUE #9584: Legends implementation
There is several work in CARTO.js and Builder. But let's tackle first the implementation (both sides, CARTO.js and Builder), API (both sides again), panel views (Builder) and so on, leaving proper legends styles (markup) for the end of the project (where @piensaenpixel will help a lot).

cc @piensaenpixel @saleiva @alonsogarciapablo

### PR DESCRIPTION
**Pull request by nobuti**
This PR fixes #9584.

Still work in progress:
- [x] Style model linked to bubble model
- [x] Random color for category item model
- [x] HTML editor: content
  - [x] Load HTML (`pre-html-snippet` and `post-html-snippet`)
  - [x] Split `pre-html-snippet` and `post-html-snippet` in two fields
  - [x] Codemirror readonly lines http://codemirror.net/doc/manual.html#markText
  - [x] Codemirror search placeholder to mark lines as readonly
  - [x] Sanitize html
- [x] Legend collections and model
- [ ] ~~Map options: add legends~~
- [x] Deep insights integrations for cartodb.js
- [x] Listen styleModel for magic legends
- [x] Make legend types equal to cartodbjs
- [x] Category form model populated with the styles
- [x] Infoboxes
- [x] Edge case: migration legend
- [ ] Backend integration (depends on #9710)
- [x] cartodb.js integration (https://github.com/CartoDB/cartodb.js/pull/1401)

cc @alonsogarciapablo @xavijam

### CODE REVIEW

File: app/assets/stylesheets/editor-3/legend-list-form.css.scss [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+.LegendItem .Editor-FormDialog {
+  left: 0;
+}
+
+.LegendItem .Editor-fieldset {
+  @include display-flex();
+  @include justify-content(space-between);
+}
+
+.LegendItem .Editor-fieldset .Editor-formInner:nth-child(1) {
```
Comment (nobuti) [inline app/assets/stylesheets/editor-3/legend-list-form.css.scss]: @piensaenpixel I hate it too, but backbone-forms doesn't allow us to provide a form inner template for nested lists :(

File: lib/assets/javascripts/cartodb3/components/form-components/editors/list/list-item.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * A single item in the list
+ *
+ * @param {editors.List} options.list The List editor instance this item belongs to
+ * @param {Function} options.Editor   Editor constructor function
+ * @param {String} options.key        Model key
+ * @param {Mixed} options.value       Value
+ * @param {Object} options.schema     Field schema
+ */
+Backbone.Form.editors.List.Item = Backbone.Form.editors.Base.extend({
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/components/form-components/editors/list/list-item.js]: This addon belongs to the backbone-forms library, but they don't provide a commonjs version.

File: lib/assets/javascripts/cartodb3/components/form-components/editors/list/list.js [hunk 3]
```diff
@@ -0,0 +1,250 @@
+var Backbone = require('backbone');
+var $ = require('jquery');
+var _ = require('underscore');
+
+Backbone.Form.editors.List = Backbone.Form.editors.Base.extend({
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/components/form-components/editors/list/list.js]: Same as above.

File: lib/assets/javascripts/cartodb3/components/form-components/editors/sortable-list.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    this._initSortable();
+    return this;
+  },
+
+  _initSortable: function () {
+    this.$list.sortable({
+      axis: 'y',
+      items: '.js-sortable',
+      tolerance: 'pointer',
+      // placeholder: 'Editor-ListLayer-item Editor-ListLayer-item--placeholder',
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/components/form-components/editors/sortable-list.js]: @piensaenpixel I need a hint here, what class(es) should I apply to it in order to get the pretty placeholder?

File: lib/assets/javascripts/cartodb3/components/form-components/editors/sortable-list.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    Backbone.Form.editors.List.prototype.render.apply(this, arguments);
+    this._initSortable();
+    return this;
+  },
+
+  _initSortable: function () {
+    this.$list.sortable({
+      axis: 'y',
+      items: '.js-sortable',
+      tolerance: 'pointer',
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/components/form-components/editors/sortable-list.js]: With this, we avoid issues when dragging elements to position zero. Maybe we could apply it to the layers view, I think there is an issue open, I couldn't reproduce it though.

File: lib/assets/javascripts/cartodb3/components/form-components/editors/text.js [hunk 6]
```diff
@@ -18,8 +18,10 @@ Backbone.Form.editors.Text = Backbone.Form.editors.Text.extend({
   _toggleDisableState: function () {
     if (this.options.disabled) {
       this.$el.attr('readonly', '');
+      this.$el.attr('placeholder', '');
     } else {
       this.$el.removeAttr('readonly');
+      this.$el.attr('placeholder', this.options.placeholder || '');
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/components/form-components/editors/text.js]: To add placeholders to the `enabler-editor` component.

File: lib/assets/javascripts/cartodb3/data/legend-definition-model.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+var Backbone = require('backbone');
+var syncAbort = require('./backbone/sync-abort');
+
+/*
+  Base model for a legend. It should have a reference to a layerDefinitionModel.
+ */
+module.exports = Backbone.Model.extend({
+  defaults: {
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/data/legend-definition-model.js]: @alonsogarciapablo @guido the naming convention for legends properties, let me know if they fit in your thoughts.

File: lib/assets/javascripts/cartodb3/deep-insights-integration/legend-manager.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
+var LegendFactory = require('../editor/layers/layer-content-views/legend/legend-factory');
+
+var onChange = function (layerDefModel) {
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/deep-insights-integration/legend-manager.js]: We listen changes in styleModel in order to _suggest_ legends to the user, even she/he didn't select any of them.

File: lib/assets/javascripts/cartodb3/editor.js [hunk 9]
```diff
@@ -153,6 +154,12 @@ deepInsights.createDashboard('#dashboard', vizJSON, {
 
   layerDefinitionsCollection.resetByLayersData(layersData);
 
+  var legendDefinitionsCollection = new LegendDefinitionsCollection(null, {
+    configModel: configModel,
+    layerDefinitionsCollection: layerDefinitionsCollection
+  });
+  legendDefinitionsCollection.resetByData(vizJSON);
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor.js]: We use the vizjson to populate the legend definitions collection, in the same way we do with layers.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/form/legend-color-range.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  '#305482',
+  '#CCD859',
+  '#565175',
+  '#FFB927',
+  '#3EBCAE',
+  '#E54C1F'
+];
+
+module.exports = {
+  getNextColor: function () {
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/form/legend-color-range.js]: When the user chooses a custom category legend, the colors added are taken from the collection above.

File: lib/assets/javascripts/cartodb3/data/legend-definition-model.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    type: 'none',
+    title: '',
+    items: [],
+    prefix: '',
+    suffix: '',
+    'raw-html': '',
+    'pre-html-snippet': '',
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/legend-definition-model.js]: `pre-html-snippet`, `pre_html_snippet`, `preHTMLSnippet` ? cc: @CartoDB/frontend

File: lib/assets/javascripts/cartodb3/data/legend-definition-model.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    'post-html-snippet': '',
+    fill: ''
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/legend-definition-model.js]: Some of these attributes look like there are specific to certain types of legends (eg: `raw-html` would only apply to HTML Legends, `fill` would only apply to bubble legends, etc.). It makes me wonder if they should be part of this base model...

Comment (matallo) [inline lib/assets/javascripts/cartodb3/data/legend-definition-model.js]: +1 to extend specific models

File: lib/assets/javascripts/cartodb3/data/legend-definition-model.js (hunk 11 above)
Comment (matallo) [inline lib/assets/javascripts/cartodb3/data/legend-definition-model.js]: +1 to `preHTMLSnippet`

**Comment by xavijam** [comment]
What do you mean with the 'infobox' point @nobuti ?

**Comment by xavijam** [comment]
And why have we removed "map options: add legends" from the scope?

**Comment by nobuti** [comment]
@xavijam implement infoboxes properly, you know `layer hidden`, `migrated legend` and so. About the map options, I've pull it out in its own issue: https://github.com/CartoDB/cartodb/issues/9751 in order to simplify a little bit the PR.

**Comment by xavijam** [comment]
Ahhhh, great, thanks man!

**Comment by xavijam** [comment]
How bad or good would it be if we integrate it without the backend thing?
I bet for creating a feature-flag, in order to activate or not the legends panel, and then when backend services are ready, enable it by default and remove the feature flag.

**Comment by nobuti** [comment]
Sounds good, it could be easily the last step, let me get it ready.

**Comment by xavijam** [comment]
\o/ thanks!

File: lib/assets/javascripts/cartodb3/components/code-mirror/searchcursor.code-mirror.js [hunk 13]
```diff
@@ -0,0 +1,183 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/code-mirror/searchcursor.code-mirror.js]: Q: couldn't we get this "addon" from the codemirror package?

File: lib/assets/javascripts/cartodb3/components/form-components/editors/list/list-item.js [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+  /*eslint-disable */
+  template: _.template('\
+      <div>\
+        <span data-editor></span>\
+        <button type="button" data-action="remove">&times;</button>\
+      </div>\
+    ', null, Backbone.Form.templateSettings),
+
+  // template: _.template('\
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/form-components/editors/list/list-item.js]: Remove this commented code.

File: lib/assets/javascripts/cartodb3/data/legends/legend-definitions-collection.js [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    if (!options.layerDefinitionsCollection) throw new Error('layerDefinitionsCollection is required');
+
+    this.configModel = options.configModel;
+    this.layerDefinitionsCollection = options.layerDefinitionsCollection;
+  },
+
+  url: function () {
+    var mapId = this.layerDefinitionModel.collection.mapId;
+    var baseUrl = this.configModel.get('base_url');
+    // TOFIX WHEN THE API IS READY
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/legends/legend-definitions-collection.js]: TODO then?

File: lib/assets/javascripts/cartodb3/deep-insights-integration/legend-manager.js (hunk 8 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/deep-insights-integration/legend-manager.js]: \o/, if it comes from a migrated one, no magic, right?

File: lib/assets/javascripts/cartodb3/editor.js (hunk 9 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor.js]: :+1:

File: lib/assets/javascripts/cartodb3/editor.js [hunk 16]
```diff
@@ -105,10 +107,10 @@ UserGroupFetcher.track({
 });
 
 // Setup and create the vis (map, layers etc.) + dashboard (widgets)
-// from the given vizJSON, but disabling legends
+// from the given vizJSON, but enabling legends behind a feature flag
 vizJSON = _.extend(
   vizJSON, {
-    legends: false
+    legends: userModel.featureEnabled('builder-legends')
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor.js]: :+1:

File: lib/assets/javascripts/cartodb3/deep-insights-integration/legend-manager.js (hunk 8 above)
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/deep-insights-integration/legend-manager.js]: Yes, two lines below we check if that layer has a migrated legend.

File: lib/assets/javascripts/cartodb3/components/code-mirror/searchcursor.code-mirror.js (hunk 13 above)
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/components/code-mirror/searchcursor.code-mirror.js]: I tried with no success 😞

**Comment by nobuti** [comment]
🚀

**Pull request merged**